File: week12/15/4_Thu_Queue_2stacks.py
class MyQueue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []
        self.size = 0

    def push(self, x: int) -> None:
        if self.stack1:
            while self.stack1:
                dat = self.stack1.pop()          
                self.stack2.append(dat)            
            self.stack1.append(x)

            while self.stack2:
                dat = self.stack2.pop()
                self.stack1.append(dat)
        else:
            self.stack1.append(x)
        self.size += 1

    def pop(self) -> int:
        if self.stack1:            
            self.size -= 1
            return self.stack1.pop()

    def peek(self) -> int:
        return self.stack1[-1]

    def empty(self) -> bool:
        if self.size == 0:
            return True
        else:
            return False
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: week12/15/4_Thu_Queue_2stacks.cpp
#include <stack>

using namespace std;

class MyQueue {
public:
    stack<int> stack1, stack2;
    MyQueue() {        
    }
    
    void push(int x) {
        if (!stack1.empty()){
            while (!stack1.empty()) {
                int dat = stack1.top();   
                stack1.pop();             
                stack2.push(dat);
            }
            stack1.push(x);
            while (!stack2.empty()) {
                int dat = stack2.top();
                stack2.pop();
                stack1.push(dat);
            }        
        }
        else stack1.push(x);
    }
    
    int pop() {
        int dat = stack1.top();
        stack1.pop();
        return dat;
    }
    
    int peek() {
        return stack1.top();
    }
    
    bool empty() {
        return stack1.empty();
        
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */